packages/xdl/src/start/ExpoUpdatesManifestHandler.ts

```diff
@@ -0,0 +1,155 @@
+import http from 'http';
+import { parse } from 'url';
+import uuid from 'uuid';
+
+import {
+  Analytics,
+  Config,
+  ProjectAssets,
+  ProjectUtils,
+  resolveEntryPoint,
+  UrlUtils,
+} from '../internal';
+import {
+  getBundleUrlAsync,
+  getExpoGoConfig,
+  getPackagerOptionsAsync,
+  stripPort,
+} from './ManifestHandler';
+
+function getPlatformFromRequest(req: express.Request | http.IncomingMessage): string {
+  const url = req.url ? parse(req.url, true) : null;
+  const platform = url?.query.platform || req.headers['expo-platform'];
+  if (!platform) {
+    throw new Error('Must specify expo-platform header or query parameter');
+  }
+  return String(platform);
+}
+
+export async function getManifestResponseAsync({
+  projectRoot,
+  platform,
+  host,
+}: {
+  projectRoot: string;
+  platform: string;
+  host?: string;
+}): Promise<{
+  body: ExpoUpdatesManifest;
+  headers: Map<string, number | string | readonly string[]>;
+}> {
+  const headers = new Map<string, any>();
+  // set required headers for expo-updates manifest specification
+  headers.set('expo-protocol-version', 0);
+  headers.set('expo-sfv-version', 0);
+  headers.set('cache-control', 'private, max-age=0');
+  headers.set('content-type', 'application/json');
+
+  const hostname = stripPort(host);
+  const [projectSettings, bundleUrlPackagerOpts] = await getPackagerOptionsAsync(projectRoot);
+  const projectConfig = getConfig(projectRoot);
+  const entryPoint = resolveEntryPoint(projectRoot, platform, projectConfig);
+  const mainModuleName = UrlUtils.stripJSExtension(entryPoint);
+  const expoConfig = projectConfig.exp;
+  const expoGoConfig = await getExpoGoConfig({
+    projectRoot,
+    projectSettings,
+    mainModuleName,
+    hostname,
+  });
+
+  const hostUri = await UrlUtils.constructHostUriAsync(projectRoot, hostname);
+
+  const runtimeVersion =
+    expoConfig.runtimeVersion ??
+    (expoConfig.sdkVersion ? getRuntimeVersionForSDKVersion(expoConfig.sdkVersion) : null);
+  if (!runtimeVersion) {
+    throw new Error('Must specify runtimeVersion or sdkVersion in app.json');
+  }
+
+  const bundleUrl = await getBundleUrlAsync({
+    projectRoot,
+    platform,
+    projectSettings,
+    bundleUrlPackagerOpts,
+    mainModuleName,
+    hostname,
+  });
+
+  // For each manifest asset (for example `icon`):
+  // - set a field on the manifest containing a reference to the asset: iconAsset: { rawUrl?: string, assetKey?: string }
+  // - gather the data needed to embed a reference to that asset in the expo-updates assets key
+  const assets = await ProjectAssets.resolveAndCollectExpoUpdatesManifestAssets(
+    projectRoot,
+    expoConfig,
+    path => bundleUrl!.match(/^https?:\/\/.*?\//)![0] + 'assets/' + path
+  );
+
+  const expoUpdatesManifest = {
+    id: uuid.v4(),
+    createdAt: new Date().toISOString(),
+    runtimeVersion,
+    launchAsset: {
+      key: mainModuleName,
+      contentType: 'application/javascript',
+      url: bundleUrl,
```

Comment:
A few assumptions here may change with Hermes: the content type won't be JS and the URL will point to an HBC file instead of a JS bundle.

Reply:
Yep, should be easy to change at that time.